zerver/tests/test_channel_permissions.py

```diff
@@ -746,3 +755,506 @@ def test_change_stream_message_retention_days_requires_realm_owner(self) -> None
         self.assert_json_success(result)
         stream = get_stream("stream_name1", realm)
         self.assertEqual(stream.message_retention_days, 2)
+
+
+class PermissionCheckConfigDict(TypedDict):
+    setting_group: NamedUserGroup | UserGroupMembersData
+    users_with_permission: list[UserProfile]
+    users_without_permission: list[UserProfile]
+
+
+class ChannelAdministerPermissionTest(ZulipTestCase):
+    @override
+    def setUp(self) -> None:
+        super().setUp()
+        self.realm = get_realm("zulip")
+        self.admin = self.example_user("iago")
+        self.moderator = self.example_user("shiva")
+        self.guest = self.example_user("polonius")
+
+        self.hamletcharacters_group = NamedUserGroup.objects.get(
+            name="hamletcharacters", realm=self.realm
+        )
+        self.moderators_group = NamedUserGroup.objects.get(
+            name=SystemGroups.MODERATORS, realm=self.realm, is_system_group=True
+        )
+        self.nobody_group = NamedUserGroup.objects.get(
+            name=SystemGroups.NOBODY, realm=self.realm, is_system_group=True
+        )
+        self.members_group = NamedUserGroup.objects.get(
+            name=SystemGroups.MEMBERS, realm=self.realm, is_system_group=True
+        )
+
+    def do_test_updating_channel(
+        self, stream: Stream, property_name: str, new_value: str | int | bool
+    ) -> None:
+        hamlet = self.example_user("hamlet")
+        prospero = self.example_user("prospero")
+
+        # For some properties, name of the field in Stream model
+        # is different from parameter name used in API request.
+        api_parameter_name_dict = dict(
+            name="new_name",
+            deactivated="is_archived",
+        )
+        api_parameter_name = property_name
+        if property_name in api_parameter_name_dict:
+            api_parameter_name = api_parameter_name_dict[property_name]
+
+        data = {}
+        if not isinstance(new_value, str):
+            data[api_parameter_name] = orjson.dumps(new_value).decode()
+        else:
+            data[api_parameter_name] = new_value
+
+        default_error_msg = "You do not have permission to administer this channel."
+
+        def check_channel_property_update(
+            user: UserProfile, allow_fail: bool = False, error_msg: str = default_error_msg
```

Comment:
`s/allow_fail/expect_fail/` would be a better name. I think you probably want just a single parameter, like `expected_error: str`, rather than two parameters that are basically duplicate. We use that sort of pattern in `test_classes` a few places I think.